trackpy/static.py

```diff
@@ -0,0 +1,343 @@
+
+    Parameters
+    ----------
+    feat : Pandas DataFrame
+        DataFrame containing the x and y coordinates of particles
+    cutoff : float
+        Maximum distance to calculate g(r)
+    fraction : float, optional
+        The fraction of particles to calculate g(r) with. May be used to
+        increase speed of function. Particles selected at random.
+    dr : float, optional
+        The bin width
+    p_indices : sequence, optional
+        Only consider a pair of particles if one of them is in 'p_indices'.
+        Uses zero-based indexing, regardless of how 'feat' is indexed.
+    ndensity : float, optional
+        Density of particle packing. If not specified, density will be
+        calculated assuming rectangular homogenous
+        arrangement.
+    boundary : tuple, optional
+        Tuple specifying rectangular prism boundary of particles (xmin, xmax,
+        ymin, ymax). Must be floats. Default is to assume a rectangular packing.
+        Boundaries are determined by edge particles.
+    handle_edge : boolean, optional
+        If true, compensate for reduced area around particles near the edges.
+    max_ndensity : number, optional
+        The relative maximum density deviation, used to estimate the maximum
+        number of neighbours. Lower numbers increase performance, until the
+        method fails because there are more neighbours than expected.
+
+    Returns
+    -------
+    r_edges : array
+        Return the bin edges
+    g_r : array
+        The values of g_r
+    """
+
+    if boundary is None:
+        xmin, xmax, ymin, ymax = (feat.x.min(), feat.x.max(),
+                                  feat.y.min(), feat.y.max())
+    else:
+        xmin, xmax, ymin, ymax = boundary
+
+        # Disregard all particles outside the bounding box
+        feat = feat[(feat.x >= xmin) & (feat.x <= xmax) &
+                    (feat.y >= ymin) & (feat.y <= ymax)]
+
+    if ndensity is None:  # particle packing density
+        ndensity = (feat.x.count() - 1) / ((xmax - xmin) * (ymax - ymin))
+
+    if p_indices is None:
+        if fraction == 1.:
+            p_indices = slice(len(feat))
+        else:  # grab random sample of particles
+            p_indices = np.random.randint(0, len(feat),
+                                          int(fraction * len(feat)))
+
+    # radii bins to search for particles
+    r_edges = np.arange(0, cutoff + dr, dr)
+
+    # initialize kdtree for fast neighbor search
+    ckdtree = cKDTree(feat[['x', 'y']])
+    pos = ckdtree.data[p_indices]
+
+    # Estimate upper bound for neighborhood particle count
+    max_p_count = int(np.pi * (r_edges.max() + dr)**2 *
+                      ndensity * max_rel_ndensity)
+    # Protect against too large memory usage
+    if len(pos) * max_p_count > MAX_ARRAY_SIZE:
+        max_p_count = int(MAX_ARRAY_SIZE / len(pos))
+
+    dist, idxs = ckdtree.query(pos, k=max_p_count, distance_upper_bound=cutoff)
+    if np.any(np.isfinite(dist[:, -1])):
```

Comment:
We should probably be more concise over here, I was confusing 2 different errors here:

At line 86, the `max_p_count` (so indirectly `max_rel_ndensity`) is reduced automatically if the produced array would be larger than `MAX_ARRAY_SIZE`. This should produce an error instead of changing the parameter silently. This error should recommend lowering either `max_rel_ndensity`, `cutoff`, or `fraction` as they all reduce the array size.

At line 90, there is a different issue (as @nkeim described): `max_rel_ndensity` is too low. The exception should say that this should be increased. (Which could lead to the first exception).